File: lc43.py
class Solution(object):
    def multiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        if num1 == '0' or num2 == '0':
            return '0'
        l1, l2 = len(num1), len(num2)
        muls = []
        p1 = l1 - 1
        maxl = 0
        while p1 >= 0:
            tmp = []
            for i in xrange(l1-1-p1):
                tmp.append('0')
            carry = 0
            n1 = int(num1[p1])
            p2 = l2 - 1
            while p2 >= 0:
                n2 = int(num2[p2])
                tmp.append(str((n1*n2 + carry) % 10))
                carry = (n1 * n2 + carry) / 10
                p2 -= 1
            p1 -= 1
            while carry != 0:
                tmp.append(str(carry % 10))
                carry /= 10
            tmp.reverse()
            maxl = max(maxl, len(tmp))
            muls.append(''.join(tmp))
        ret = []
        carry = 0
        for i in xrange(-1, -maxl-1, -1):
            s = 0
            for res in muls:
                if abs(i) <= len(res):
                    s += int(res[i])
            ret.append(str((s + carry) % 10))
            carry = (s + carry) / 10
        while carry != 0:
            ret.append(str(carry % 10))
            carry /= 10
        ret.reverse()
        return ''.join(ret)

File: lc43.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        if (num1 == "0" || num2 == "0") return "0";
        int l1 = num1.size(), l2 = num2.size();
        int p1 = l1 - 1, p2;
        int carry = 0, maxlen = 0;
        vector<string> toBeAdded;
        while (p1 >= 0) {
            stringstream ss("");
            p2 = l2 - 1;
            for (int i = p1; i < l1 - 1; ++i) ss << '0';
            carry = 0;
            while (p2 >= 0) {
                int tmp = (num1[p1] - '0') * (num2[p2] - '0') + carry;
                ss << (tmp % 10);
                carry = tmp / 10;
                --p2;
            }
            if (carry > 0) ss << to_string(carry);
            string midProd(ss.str());
            reverse(midProd.begin(), midProd.end());
            toBeAdded.emplace_back(midProd);
            --p1;
        }
        map_add(toBeAdded);
        return toBeAdded[0];
    }
private:
    void map_add(vector<string>& nums) {
        while (nums.size() >= 2) {
            string num1 = nums.back();
            nums.pop_back();
            string num2 = nums.back();
            nums.pop_back();
            nums.emplace_back(string_add(num1, num2));
        }
    }
    
    string string_add(string num1, string num2) {
        stringstream ss("");
        int p1 = num1.size() - 1, p2 = num2.size() - 1, carry = 0;
        while (p1 >= 0 && p2 >= 0) {
            int sum = (num1[p1] - '0') + (num2[p2] - '0') + carry;
            ss << sum % 10;
            carry = sum / 10;
            --p1;
            --p2;
        }
        while (p1 >= 0) {
            int sum = num1[p1] - '0' + carry;
            ss << sum % 10;
            carry = sum / 10;
            --p1;
        }
        while (p2 >= 0) {
            int sum = num2[p2] - '0' + carry;
            ss << sum % 10;
            carry = sum / 10;
            --p2;
        }
        if (carry > 0) ss << to_string(carry);
        string res(ss.str());
        reverse(res.begin(), res.end());
        return res;
    }
};